### ISSUE TITLE
More intuitive transforming between content blocks

### ISSUE THREAD

**Issue by richtabor** [WordPress/gutenberg#55067]
It's cumbersome switching between text blocks in the block editor. 

The primary method is to select the block icon from the toolbar, then transform to another block, then apply secondary "transforms" (i.e. heading level) if necessary. 

**We can combine these varying flows into one**, while also reducing the cognitive load on transforming between varying text elements (blocks). I know changing from a H1 to H2 is _technically_ not a transform as-is, but it actuality it is—it's no different than transforming from any heading to a paragraph. 

I think this effort would help push the editor towards a world-class writing experience.

### Proposal
- Add heading block variations (not directly insertable in the inserter, only the "Heading" block should be insertable, like it is today, with a H2 level). 
- Add transforms to specific heading block variations from the paragraph block. This would allow for level-specific icons in the toolbar, providing a central affordance for indicating what heading this is. 
- Remove the HeadingLevelDropdown as it's no longer needed. 
- Omit content-based blocks (paragraph, heading, quote, pull quote, list) from the auto-opt-in columns and group transforms. 
- Omit content-based blocks from the "Patterns" menu item within the transforms UI. Maybe "Styles" as well.
- Consider reducing the paragraph transforms that don't cover the majority of use cases (paragraph to button, reformatted, verse, etc).  

https://github.com/WordPress/gutenberg/assets/1813435/07a4b972-ed78-416a-8c9b-b01f5aec20a9

### Current

Switching from a paragraph and to a specific heading (using the toolbar) requires first clicking on the paragraph icon (to transform), then select the heading block from the options, then selecting the heading level toolbar control. 

Switching from a heading to a paragraph requires less effort, but you can't do it where you're expecting (the level control). Also, the heading bookmark icon doesnt tell you much about this content — the "H{x}" icon does.

https://github.com/WordPress/gutenberg/assets/1813435/7b01aba0-915b-4a16-b558-62bd1f0fe9d0

**Comment by richtabor** [comment]
> Add heading block variations (not directly insertable in the inserter, only the "Heading" block should be insertable, like it is today, with a H2 level).

We should still have a "Heading" block in the inserter. Perhaps on insertion the level (variation) is set to "H2", emulating what we have today.

**Comment by SaxonF** [comment]
Makes sense to me. We need to find a balance between the editor feeling like a design tool vs a writing tool. This closely resembles writing tools but not to the extent that it negatively impacts "designing" in the editor.

**Comment by ndiego** [comment]
Functionally, I like this approach. I do think it will be one of those technically small changes but one that will have a big impact on users who are comfortable with the current setup. 

> Omit content-based blocks (paragraph, heading, quote, pull quote, list) from the auto-opt-in columns and group transforms.

Currently, these transforms place the content in the containers on transform, so something to consider.

**Comment by richtabor** [comment]
> Currently, these transforms place the content in the containers on transform, so something to consider.

We do have "Group" in the block options panel though. Not one for columns, but I think that's fine.

**Comment by ndiego** [comment]
> We do have "Group" in the block options panel though. Not one for columns, but I think that's fine.

Yeah more often than not, it feels like unexpected behavior IMO. The transform should _transform_ the block, not wrap it in another block.

**Comment by priethor** [comment]
Most of these changes make sense to me, too. More often than not, I find myself converting from paragraphs to headings and then selecting the heading level. After a few ones, this becomes a friction point.

> Yeah more often than not, it feels like unexpected behavior IMO. The transform should transform the block, not wrap it in another block.

I don't have strong opinions, but I am not so sure about this; while technically correct, if we want to prioritize the _flow_ we can allow some wiggle room on what a transform does, as long as it reduces friction and improves the writing experience.

**Comment by bgardner** [comment]
I really like this and it will definitely save me an arduous step during the transformation process.

**Comment by mtias** [comment]
Implemented levels as variations which handle a lot of these ideas by default in #73823
